setuptools/validation.py

```diff
@@ -0,0 +1,106 @@
+"""
+Validation logic for package metadata, supporting PEP 566.
+See: https://www.python.org/dev/peps/pep-0566/
+"""
+
+
+def single_line_str(val):
+    """A string that does not contain newline characters"""
+    if val is None:
+        return
+    assert isinstance(val, str), \
+        "is of type %s but should be of type str" % type(val)
+    assert '\n' not in val, \
+        "to contain a newline character (not allowed)"
+
+
+def multi_line_str(val):
+    """A string that can contain newline characters"""
+    if val is None:
+        return
+    assert isinstance(val, str), \
+        "is of type %s but should be of type str" % type(val)
+
+
+def list_of_str(val):
+    """A list of single-line text strings"""
+    if val is None:
+        return
+    assert isinstance(val, list), \
+        "is of type %s but should be of type list(str)" % type(val)
+    for item in val:
+        single_line_str(item)
+
+
+def set_of_str(val):
+    """A set (i.e. a list of unique entries) of single-line text strings"""
+    if val is None:
+        return
+    assert isinstance(val, set), \
+        "is of type %s but should be of type set(str)" % type(val)
+    for item in val:
+        single_line_str(item)
+
+
+default_validator = single_line_str
+validators = dict(
+    long_description=multi_line_str,
+    classifiers=list_of_str,
+    keywords=list_of_str,
+    platforms=list_of_str,
+    project_urls=list_of_str,
+    requires=list_of_str,
+    provides=list_of_str,
+    provides_extras=set_of_str,
+    obsoletes=list_of_str,
+)
+
+
+class Metadata:
+    """
+    A validation object for PKG-INFO data fields
+    """
+
+    def __init__(self, *data_dicts, **kwargs):
+        """
+        Takes JSON-compatible metadata, as described in PEP 566, which
+        can be added as both dictionaries or keyword arguments.
+        """
+        self.errors = []
+        self.fields = []
+
+        for data in data_dicts:
+            for key in data:
+                setattr(self, key, data[key])
+                self.fields += [key]
+
+        for key in kwargs:
+            setattr(self, key, kwargs[key])
+            self.fields += [key]
+
+    def validate(self, throw_exception=False):
```

Comment:
Please note that @di [suggested to add validation functionality](https://github.com/pypa/setuptools/issues/1390#issuecomment-432411556) that behaves like `validate().errors`, so that it integrates nicely into Warehouse code.

I felt that throwing an exception by default would make code more clumsy. It's just not beautiful. -- I wouldn't want to miss that possibility in general, though.

Reply:
> Please note that @di suggested to add validation functionality that behaves like validate().errors, so that it integrates nicely into Warehouse code.

He suggested no such thing. He suggested that something like the warehouse functionality go into a *different* project, [pypa/packaging](https://github.com/pypa/packaging), not this one. I think you may have misunderstood him.

> I felt that throwing an exception by default would make code more clumsy. It's just not beautiful. -- I wouldn't want to miss that possibility in general, though.

This is in general not an example of good design, because there are now two different error handling pathways. Additionally, the Pythonic way to signal an error condition is to raise an exception. It is not clumsy and in fact it leads to a lot of clumsy code in downstream consumers. The reason is that in most cases, the functions calling your function can do nothing about an error except abort early and notify the user, if you raise an exception in error conditions, anyone who isn't going to try to recover from the error can just use your function as if it always works because the exception you raise will bubble up the stack until someone either catches it or the program is aborted and the user is notified of the error. If you return an error code instead, you are relying on your consumers to know that your function can either return an error code of some sort or not and raise an error or a handler.